### PR TITLE
Add fused linear cross entropy CPU implementation

### PR DESCRIPTION
**Pull request by krocki**
ghstack version of https://github.com/pytorch/pytorch/pull/163774

Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.12.0) (oldest at bottom):
* #169460
* #169600
* __->__ #169459

### CODE REVIEW

File: aten/src/ATen/native/native_functions.yaml [hunk 1]
```diff
@@ -9614,6 +9614,30 @@
   dispatch:
     CompositeImplicitAutograd: cross_entropy_loss_symint
 
+- func: linear_cross_entropy(Tensor self, Tensor linear_weight, Tensor target, *, Tensor? linear_bias=None, int reduction=Mean, SymInt ignore_index=-100, float label_smoothing=0.0, str chunking_strategy="none", int? vocab_chunk_size=None, int? batch_chunk_size=None) -> Tensor
```
Comment (mikaylagawarecki) [inline aten/src/ATen/native/native_functions.yaml:9617]: Why do we need to register this version of the function to native_functions.yaml? If I understand correctly, this is intended to be for inference as it passes `save_for_backward=False` to the underlying impl, is that correct?

I have a follow up question here -- is an example of the case where you expect this to be used to get log_probs (e.g. for GRPO and such)?

Comment (mikaylagawarecki) [reply aten/src/ATen/native/native_functions.yaml:9617]: I'm trying to understand the nuance in supporting the inference path a bit better, can you clarify whether my understanding is correct -- 

Other ops (e.g. `fused_rms_norm` that additionally returns `rstd`) also have a convention where they save extra tensors for backward. 

When gradients are not enabled, these intermediates are still computed and returned, but autograd doesn't hold references to them. So they get garbage collected immediately. However, computing `rstd` is cheap and happens regardless, so it's ok to always compute it.

On the other hand, in this case, what we're computing and saving for backward in forward is additional computation, so it is expensive and we need a separate inference path

Could you confirm that my understanding is correct?

Comment (mikaylagawarecki) [reply aten/src/ATen/native/native_functions.yaml:9617]: If all the above is correct, I might recommend that we expose `save_for_backward` directly in 

`_linear_cross_entropy_{vocab/batch}_chunking` (you can use Tensor? to indicate optional returns in the native_functions signature)

and then in python we can do something like

```
# in torch/nn/functional.py
def linear_cross_entropy(...):
   if not torch.is_grad_enabled():
       # call relevant functions with save_for_backward=False
   else:
       # call relevant functions with save_for_backward=True
```

Maybe cc @soulitzer to confirm that this recommendation sound reasonable

File: test/test_decomp.py [hunk 2]
```diff
@@ -1318,9 +1333,9 @@ def all_aten_overloads():
         }
         ops_missing_decomp = overloads_wanting_decomp - decomposition_table.keys()
         ops_missing_decomp -= allow_list
-        self.assertExpected(
-            "".join(sorted(op.name() + "\n" for op in ops_missing_decomp))
-        )
+        filtered_ops = _drop_linear_ce_ops(ops_missing_decomp)
```
Comment (mikaylagawarecki) [inline test/test_decomp.py:1336]: why did this change have to made here, isn't there a skiplist somewhere?

File: torch/nn/functional.py [hunk 3]
```diff
@@ -3512,6 +3512,183 @@ def cross_entropy(
...
+        raise ValueError(
+            f"batch_chunk_size must be positive when provided, got {batch_chunk_size}"
+        )
+
+    if (
+        input.device.type != "cpu"
+        or linear_weight.device != input.device
+        or target.device != input.device
+        or (linear_bias is not None and linear_bias.device != input.device)
+    ):
```
Comment (mikaylagawarecki) [inline torch/nn/functional.py]: Same comment from my previous review https://github.com/pytorch/pytorch/pull/163774#discussion_r2519874166

> it's ok for the dispatcher to throw an unhandled error if the implementation has not been registered to the corresponding backend key (e.g. CUDA etc.) yet.
> Per below comments, this does not need to fall back to unfused for other devices, also please document the arguments like how other docstrings in this file are written

**Review by mikaylagawarecki — COMMENTED** [review]
I have some clarifying questions re the separate API that seems to be for inference?

Separately, is my understanding correct that the PR stacked directly above fixes the test failures in this PR?

File: aten/src/ATen/native/native_functions.yaml (hunk 1 above)
Comment (soulitzer) [reply aten/src/ATen/native/native_functions.yaml:9617]: Yeah that sounds reasonable
